daemon/src/collab_settlement_taker.rs

```diff
@@ -0,0 +1,194 @@
+use crate::address_map::Stopping;
+use crate::model::cfd::{
+    Cfd, CollaborativeSettlement, OrderId, SettlementKind, SettlementProposal,
+};
+use crate::model::Price;
+use crate::{connection, projection, wire};
+use anyhow::{Context, Result};
+use async_trait::async_trait;
+use xtra::prelude::MessageChannel;
+use xtra_productivity::xtra_productivity;
+
+pub struct Actor {
+    cfd: Cfd,
+    projection: xtra::Address<projection::Actor>,
+    on_completed: Box<dyn MessageChannel<Completed>>,
+    connection: xtra::Address<connection::Actor>,
+    proposal: SettlementProposal,
+}
+
+impl Actor {
+    pub fn new(
+        cfd: Cfd,
+        projection: xtra::Address<projection::Actor>,
+        on_completed: impl MessageChannel<Completed> + 'static,
+        current_price: Price,
+        connection: xtra::Address<connection::Actor>,
+        n_payouts: usize,
+    ) -> Result<Self> {
+        let proposal = cfd.calculate_settlement(current_price, n_payouts)?;
+
+        Ok(Self {
+            cfd,
+            projection,
+            on_completed: Box::new(on_completed),
+            connection,
+            proposal,
+        })
+    }
+
+    async fn propose(&mut self, this: xtra::Address<Self>) -> Result<()> {
+        if !self.cfd.is_collaborative_settle_possible() {
+            anyhow::bail!(
+                "Settlement proposal not possible because for cfd {} is in state {} which cannot be collaboratively settled",
+                self.cfd.order.id,
+                self.cfd.state
+            )
+        }
+
+        self.connection
+            .send(connection::ProposeSettlement {
+                timestamp: self.proposal.timestamp,
+                taker: self.proposal.taker,
+                maker: self.proposal.maker,
+                price: self.proposal.price,
+                address: this,
+                order_id: self.cfd.order.id,
+            })
+            .await??;
+
+        self.update_proposal(Some((self.proposal.clone(), SettlementKind::Outgoing)))
+            .await?;
+
+        Ok(())
+    }
+
+    async fn handle_confirmed(&mut self) -> Result<CollaborativeSettlement> {
+        let order_id = self.cfd.order.id;
+
+        tracing::info!(%order_id, "Settlement proposal got accepted");
+
+        self.update_proposal(None).await?;
```

Comment:
:question: Is this supposed to take `None` as an argument here. You do the same for handling a rejection.

Reply:
Yes, because in any case, we proposal is now "complete".